src/server/mod.rs

```diff
@@ -79,44 +75,40 @@ fn replace_string_in_range(
     contents
 }
 
-fn function_defines(
-    name: &str,
-    params: &[FunctionParameter],
-) -> bool {
-    params.iter().any(|param| param.key.name == name)
-}
-
-fn is_scope(name: &Symbol, n: walk::Node<'_>) -> bool {
-    let mut dvisitor =
-        semantic::DefinitionFinderVisitor::new(name.clone());
-    walk::walk(&mut dvisitor, n);
-
-    dvisitor.node.is_some()
-}
-
-fn find_references(
+fn find_references<'a>(
     uri: &lsp::Url,
-    result: NodeFinderResult,
+    node: Option<flux::semantic::walk::Node<'a>>,
+    path: Vec<flux::semantic::walk::Node<'a>>,
 ) -> Vec<lsp::Location> {
-    if let Some(node) = result.node {
+    if let Some(node) = node {
         let name = match node {
             walk::Node::Identifier(ident) => &ident.name,
             walk::Node::IdentifierExpr(ident) => &ident.name,
             _ => return Vec::new(),
         };
 
-        let mut path_iter = result.path.iter().rev();
+        let mut path_iter = path.iter().rev();
         let scope: walk::Node =
             match path_iter.find_map(|n| match n {
                 walk::Node::FunctionExpr(f)
-                    if function_defines(name, &f.params) =>
+                    if f.params
+                        .iter()
+                        .any(|param| &param.key.name == name) =>
                 {
                     Some(n)
                 }
-                walk::Node::Package(_) | walk::Node::File(_)
-                    if is_scope(name, *n) =>
-                {
-                    Some(n)
+                walk::Node::Package(_) | walk::Node::File(_) => {
+                    let mut visitor =
+                        semantic::DefinitionFinderVisitor::new(
+                            name.clone(),
+                        );
+                    walk::walk(&mut visitor, *n);
```

Comment:
Confirming what I think is going on:
* This is called when we already know the symbol (name) to lookup.
  * called per 1 symbol 
  * define a single/new `semantic::DefinitionFinderVisitor`.
* Start walk at the subtree roots (a package or file).
  * the 3 callers (rename, document_highlight, and references) all start at the root.
* The leaves:
  * a.k.a. VariableAssgn, BuiltinStmt, FunctionParameter.
  *  Stop the walk, and adds the found symbol if matches.

.

If this^^ is correct, then I don't understand what happens with variable shadowing:
* the namespace lookup starts in global scope, and then goes down the subtree.
* what if I am in an inner scope, and lookup a symbol (`find_references`) from the inner scope?
* Flux allows for shadowing: https://docs.influxdata.com/flux/v0.x/spec/assignment-scope/#variable-assignment. So think I am missing something here. 🤷🏼

Reply:
I _suspect_ when we shadow variables, we get some bug. I'm not really in a place to verify that, but it wouldn't surprise me. Most of our functions don't do a very good job of taking local scope into account (see #98 #452). As this is strictly a refactor, I'm prioritizing readability here. If you do find a bug, please file an issue about it.

Reply:
Tested, wrote issue ticket:
https://github.com/influxdata/flux-lsp/issues/479